text/0000-freeze.md

```diff
@@ -0,0 +1,155 @@
+- Feature Name: `freeze`
+- Start Date: 2020-06-12
+- RFC PR: [rust-lang/rfcs#2944](https://github.com/rust-lang/rfcs/pull/2944)
+- Rust Issue: TBD
+
+# Summary
+[summary]: #summary
+
+This RFC introduces new APIs to libcore/libstd to serve as safe abstractions for data which has no "shallow" interior mutability.
+
+```rust
+pub unsafe auto trait Freeze {}
+pub struct PhantomUnfrozen;
+```
+
+# Motivation
+[motivation]: #motivation
+
+It is occasionally necessary in systems programming to know whether the range of bytes occupied by a value is truly immutable. Given that rust has interior mutability, there is currently no way to represent this immutability in the type system.
+
+## Read Only Memory
+
+If a type is suitable for read only memory, then it cannot have any interior mutability. For example, an `AtomicU8` is a poor candidate for being put into read only memory because the type system has no way to ensure that type is not mutated. It is, however, allowed to put a `Box<AtomicUsize>` in read only memory as long as the heap allocation remains in writable memory.
```

Comment:
I found the description of `AtomicU8` somewhat confusing. I would say something like "an `AtomicU8` cannot be placed in read-only memory, since it's possible to modify it via `.store` using only a shared reference).

Reply:
Thanks for the feedback. I have pushed up the less confusing wording.